clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,18 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        KafkaChannel closingChannel = closingChannels.remove(connectionId);
+        if (closingChannel != null)
+            doClose(closingChannel, true);
```

Comment:
The reason I suggested that is that I thought we could then simplify the code here. I thought we could call `close(closingChannel, false, true)`, but looking closer, I think I was wrong. If something is in `closingChannels`, then we have already called `disconnect`. So, maybe this idea doesn't gain us anything, just an additional private method that doesn't get used by anything.